File: flutter_app/lib/core/widgets/app_text_field.dart
import 'package:flutter/material.dart';
import '../theme/design_tokens.dart';

class AppTextField extends StatefulWidget {
  final String? label;
  final String? hintText;
  final TextEditingController? controller;
  final String? initialValue;
  final bool obscureText;
  final TextInputType keyboardType;
  final String? Function(String?)? validator;
  final void Function(String)? onChanged;
  final void Function(String)? onSubmitted;
  final bool isRequired;
  final String? helperText;
  final String? errorText;
  final IconData? prefixIcon;
  final Widget? suffixIcon;
  final int? minLines;
  final int? maxLines;
  final int? maxLength;
  final bool enabled;
  final bool expands;

  const AppTextField({
    super.key,
    required this.label,
    this.controller,
    this.initialValue,
    this.hintText,
    this.obscureText = false,
    this.keyboardType = TextInputType.text,
    this.maxLines = 1,
    this.validator,
    this.onChanged,
    this.onSubmitted,
    this.prefixIcon,
    this.suffixIcon,
    this.isRequired = false,
    this.helperText,
    this.errorText,
    this.minLines,
    this.maxLength,
    this.enabled = true,
    this.expands = false,
  });

  @override
  State<AppTextField> createState() => _AppTextFieldState();
}

class _AppTextFieldState extends State<AppTextField> {
  bool _obscureText = false;

  @override
  void initState() {
    super.initState();
    _obscureText = widget.obscureText;
  }

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final colorScheme = theme.colorScheme;
    final hasError = widget.errorText != null;

    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        TextFormField(
          controller: widget.controller,
          initialValue: widget.initialValue,
          obscureText: _obscureText,
          keyboardType: widget.keyboardType,
          validator: widget.validator,
          onChanged: widget.onChanged,
          onFieldSubmitted: widget.onSubmitted,
          minLines: widget.minLines,
          maxLines: widget.obscureText ? 1 : widget.maxLines,
          maxLength: widget.maxLength,
          enabled: widget.enabled,
          expands: widget.expands,
          decoration: InputDecoration(
            // Floating label with required indicator
            label: widget.label != null
                ? Row(
                    mainAxisSize: MainAxisSize.min,
                    children: [
                      Text(widget.label!),
                      if (widget.isRequired)
                        Text(' *', style: TextStyle(color: colorScheme.error)),
                    ],
                  )
                : null,
            hintText: widget.hintText,
            helperText: widget.helperText,
            errorText: widget.errorText,
            // Leading icon with error color
            prefixIcon: widget.prefixIcon != null
                ? Icon(
                    widget.prefixIcon,
                    color: hasError
                        ? colorScheme.error
                        : colorScheme.onSurfaceVariant,
                  )
                : null,
            // Trailing icon (password toggle or custom icon or error icon)
            suffixIcon: hasError
                ? Icon(Icons.error, color: colorScheme.error)
                : widget.obscureText
                ? IconButton(
                    icon: Icon(
                      _obscureText ? Icons.visibility : Icons.visibility_off,
                      color: colorScheme.onSurfaceVariant,
                    ),
                    onPressed: () {
                      setState(() {
                        _obscureText = !_obscureText;
                      });
                    },
                  )
                : widget.suffixIcon,
            filled: true,
            fillColor: widget.enabled
                ? colorScheme.surfaceContainer
                : colorScheme.surfaceContainer.withValues(alpha: 0.5),
            // WhatsApp-style borders
            border: OutlineInputBorder(
              borderRadius: AppRadius.input,
              borderSide: BorderSide(color: colorScheme.outline),
            ),
            enabledBorder: OutlineInputBorder(
              borderRadius: AppRadius.input,
              borderSide: BorderSide(color: colorScheme.outline),
            ),
            focusedBorder: OutlineInputBorder(
              borderRadius: AppRadius.input,
              borderSide: BorderSide(color: colorScheme.primary, width: 2),
            ),
            errorBorder: OutlineInputBorder(
              borderRadius: AppRadius.input,
              borderSide: BorderSide(color: colorScheme.error),
            ),
            focusedErrorBorder: OutlineInputBorder(
              borderRadius: AppRadius.input,
              borderSide: BorderSide(color: colorScheme.error, width: 2),
            ),
            disabledBorder: OutlineInputBorder(
              borderRadius: AppRadius.input,
              borderSide: BorderSide(
                color: colorScheme.outline.withValues(alpha: 0.5),
              ),
            ),
            contentPadding: EdgeInsets.symmetric(
              horizontal: AppSpacing.md,
              vertical: AppSpacing.md,
            ),
          ),
        ),
      ],
    );
  }
}

/// TekTech AppTextArea Component
///
/// Multi-line text field for notes, descriptions, etc.
/// - Based on AppTextField with multiline support
/// - Min/max lines control
/// - Same validation and styling as AppTextField
class AppTextArea extends StatelessWidget {
  final String label;
  final TextEditingController? controller;
  final String? helperText;
  final String? Function(String?)? validator;
  final Function(String)? onChanged;
  final bool isRequired;
  final bool enabled;
  final int minLines;
  final int maxLines;
  final int? maxLength;

  const AppTextArea({
    super.key,
    required this.label,
    this.controller,
    this.helperText,
    this.validator,
    this.onChanged,
    this.isRequired = false,
    this.enabled = true,
    this.minLines = 3,
    this.maxLines = 5,
    this.maxLength,
  });

  @override
  Widget build(BuildContext context) {
    return AppTextField(
      label: label,
      controller: controller,
      helperText: helperText,
      validator: validator,
      onChanged: onChanged,
      isRequired: isRequired,
      enabled: enabled,
      minLines: minLines,
      maxLines: maxLines,
      maxLength: maxLength,
      keyboardType: TextInputType.multiline,
    );
  }
}

File: frontend/android-native/app/src/main/java/com/example/minitasktracker/core/ui/components/AppTextField.kt
package com.example.minitasktracker.core.ui.components

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Error
import androidx.compose.material.icons.filled.Visibility
import androidx.compose.material.icons.filled.VisibilityOff
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.text.input.VisualTransformation
import androidx.compose.ui.unit.dp
import com.example.minitasktracker.core.ui.theme.Radius
import com.example.minitasktracker.core.ui.theme.Spacing

/**
 * TekTech AppTextField Component
 * 
 * Modern text input field with:
 * - Label (floating)
 * - Helper text
 * - Error state with icon + message
 * - Required/optional indicator
 * - Leading/trailing icons
 * - Password visibility toggle
 * - Validation support
 * 
 * Accessibility: WCAG AA compliant, semantic labels
 * 
 * @param value Current text value
 * @param onValueChange Value change callback
 * @param label Field label
 * @param modifier Modifier
 * @param placeholder Optional placeholder text
 * @param helperText Optional helper text (shown below field)
 * @param errorText Optional error message (overrides helper)
 * @param isRequired Show required indicator (*)
 * @param enabled Whether field is enabled
 * @param readOnly Whether field is read-only
 * @param singleLine Single line input
 * @param maxLines Maximum lines for multiline
 * @param leadingIcon Optional leading icon
 * @param trailingIcon Optional trailing icon
 * @param isPassword Password field (shows visibility toggle)
 * @param keyboardOptions Keyboard options
 * @param keyboardActions Keyboard actions
 */
@Composable
fun AppTextField(
  value: String,
  onValueChange: (String) -> Unit,
  label: String,
  modifier: Modifier = Modifier,
  placeholder: String? = null,
  helperText: String? = null,
  errorText: String? = null,
  isRequired: Boolean = false,
  enabled: Boolean = true,
  readOnly: Boolean = false,
  singleLine: Boolean = true,
  maxLines: Int = 1,
  leadingIcon: ImageVector? = null,
  trailingIcon: ImageVector? = null,
  isPassword: Boolean = false,
  keyboardOptions: KeyboardOptions = KeyboardOptions.Default,
  keyboardActions: KeyboardActions = KeyboardActions.Default
) {
  var passwordVisible by remember { mutableStateOf(false) }
  val isError = errorText != null
  
  Column(modifier = modifier) {
    OutlinedTextField(
      value = value,
      onValueChange = onValueChange,
      label = {
        Row {
          Text(label)
          if (isRequired) {
            Text(
              text = " *",
              color = MaterialTheme.colorScheme.error
            )
          }
        }
      },
      placeholder = placeholder?.let { { Text(it) } },
      leadingIcon = leadingIcon?.let {
        {
          Icon(
            imageVector = it,
            contentDescription = null,
            tint = if (isError) MaterialTheme.colorScheme.error else MaterialTheme.colorScheme.onSurfaceVariant
          )
        }
      },
      trailingIcon = {
        when {
          isError -> {
            Icon(
              imageVector = Icons.Default.Error,
              contentDescription = "Error",
              tint = MaterialTheme.colorScheme.error
            )
          }
          isPassword -> {
            IconButton(onClick = { passwordVisible = !passwordVisible }) {
              Icon(
                imageVector = if (passwordVisible) Icons.Default.Visibility else Icons.Default.VisibilityOff,
                contentDescription = if (passwordVisible) "Hide password" else "Show password"
              )
            }
          }
          trailingIcon != null -> {
            Icon(
              imageVector = trailingIcon,
              contentDescription = null
            )
          }
        }
      },
      visualTransformation = if (isPassword && !passwordVisible) {
        PasswordVisualTransformation()
      } else {
        VisualTransformation.None
      },
      isError = isError,
      enabled = enabled,
      readOnly = readOnly,
      singleLine = singleLine,
      maxLines = maxLines,
      keyboardOptions = keyboardOptions,
      keyboardActions = keyboardActions,
      shape = MaterialTheme.shapes.medium,
      colors = OutlinedTextFieldDefaults.colors(
        focusedBorderColor = MaterialTheme.colorScheme.primary,
        unfocusedBorderColor = MaterialTheme.colorScheme.outline,
        errorBorderColor = MaterialTheme.colorScheme.error,
        disabledBorderColor = MaterialTheme.colorScheme.outline.copy(alpha = 0.38f)
      ),
      modifier = Modifier.fillMaxWidth()
    )
    
    // Helper or error text
    if (errorText != null || helperText != null) {
      Spacer(modifier = Modifier.height(Spacing.spacing4))
      Text(
        text = errorText ?: helperText ?: "",
        style = MaterialTheme.typography.bodySmall,
        color = if (isError) {
          MaterialTheme.colorScheme.error
        } else {
          MaterialTheme.colorScheme.onSurfaceVariant
        },
        modifier = Modifier.padding(start = Spacing.spacing16)
      )
    }
  }
}

/**
 * AppTextField variant for multiline text (notes, descriptions)
 */
@Composable
fun AppTextArea(
  value: String,
  onValueChange: (String) -> Unit,
  label: String,
  modifier: Modifier = Modifier,
  placeholder: String? = null,
  helperText: String? = null,
  errorText: String? = null,
  isRequired: Boolean = false,
  enabled: Boolean = true,
  minLines: Int = 3,
  maxLines: Int = 5,
  keyboardOptions: KeyboardOptions = KeyboardOptions.Default,
  keyboardActions: KeyboardActions = KeyboardActions.Default
) {
  AppTextField(
    value = value,
    onValueChange = onValueChange,
    label = label,
    placeholder = placeholder,
    helperText = helperText,
    errorText = errorText,
    isRequired = isRequired,
    enabled = enabled,
    singleLine = false,
    maxLines = maxLines,
    keyboardOptions = keyboardOptions,
    keyboardActions = keyboardActions,
    modifier = modifier.defaultMinSize(minHeight = (minLines * 24).dp)
  )
}
